### DISCUSSION TITLE
Could I get access to mkSynFunMatchLambdas via FCS API?

### DISCUSSION THREAD

**Discussion by 7sharp9** [dotnet/fsharp#13129] [Ideas]
**Is your feature request related to a problem? Please describe.**

During AST construction using the compiler lambda construction makes it easier to correctly define rather than duplicating info in the ast you can do a single ast construction call.

**Describe the solution you'd like**

The function `mkSynFunMatchLambdas` is callable from FCS:  
https://github.com/dotnet/fsharp/blob/2de40458360778d674cd1fd87734fbf9494ec830/src/fsharp/SyntaxTreeOps.fs#L449

**Describe alternatives you've considered**

Internal reflection, _doing this now, dont tell anyone!_

**Additional context**

There are duplicate pieces to info in the `SynExpr.Lamba` patterns between the simple patterns that are generated by the compiler via the list of patterns grabbed from the parser:

```
Lambda
                   (false, false,
                    SimplePats
                      ([Typed
                          (Id
                             (_arg1, None, true, false, false,
                              tmp.fsx (1,6--1,7)),
                           LongIdent (LongIdentWithDots ([SingleCaseDU], [])),
                           tmp.fsx (1,6--1,21))], tmp.fsx (1,5--1,22)),
                    Lambda
                      (false, true,
                       SimplePats
                         ([Typed
                             (Id
                                (value, None, false, false, false,
                                 tmp.fsx (1,24--1,29)),
                              LongIdent (LongIdentWithDots ([int], [])),
                              tmp.fsx (1,24--1,34))], tmp.fsx (1,23--1,35)),
                       App
                         (NonAtomic, false, Ident Single, Ident value,
                          tmp.fsx (1,39--1,51)), None, tmp.fsx (1,1--1,51),
                       { ArrowRange = Some tmp.fsx (1,36--1,38) }),
                    Some
                      ([Paren
                          (Typed
                             (Wild tmp.fsx (1,6--1,7),
                              LongIdent (LongIdentWithDots ([SingleCaseDU], [])),
                              tmp.fsx (1,6--1,21)), tmp.fsx (1,5--1,22));
                        Paren
                          (Typed
                             (Named (value, false, None, tmp.fsx (1,24--1,29)),
                              LongIdent (LongIdentWithDots ([int], [])),
                              tmp.fsx (1,24--1,34)), tmp.fsx (1,23--1,35))],
                       App
                         (NonAtomic, false, Ident Single, Ident value,
                          tmp.fsx (1,39--1,51))), tmp.fsx (1,1--1,51),
                    { ArrowRange = Some tmp.fsx (1,36--1,38) })
```
In this lambda snippet you can see the App section is duplicated in both the Some (pays, rhs) section, also the pats appear in the SimplePats section above.  Using the compiler  function you can just supply the pats in list form and the rhs and it will construct the nested lambda correctly.  Doing this manually is ok but prone to error.

**Comment by 7sharp9** [comment]
If a point to surface this call can be suggested then I can make the PR, trying to find where something similar is being surfaced...  Any ideas @nojaf

**Comment by 7sharp9** [comment]
Oh and if this is approved for implementation, @dsyme , Im getting ahead of myself :-)

**Comment by nojaf** [comment]
So, what is your exact use case for this?
If all you try to do is generated code by passing an AST to Fantomas, you could approach this from a slightly different angle.

Fantomas uses the `parsedData` from [SynExpr.Lambda](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synexpr.html#Lambda).
Providing just that might do the trick to getting to your code.

Something like:
```fsharp
let mkSynExprLambda (pats: SynPat list) (bodyExpr: SynExpr) =
    SynExpr.Lambda(false, false, SynSimplePats.SimplePats([], zeroRange), unitExpr, Some(pats, bodyExpr), zeroRange, { ArrowRange = Some zeroRange })
```
this shape doesn't totally add up and is not what the parser creates.
But it does contain the bits that Fantomas will use to recreate the lambda.

**Comment by 7sharp9** [comment]
The use case is only supplying the information needed to create the lambda.

Im just recreating what the parser does in that the `SynSimplePats` are generated from the `SynPat list`, theres also a duplicate of the lambda body that is present in the optional `(pats: SynPat list, rhs: SynExpr)` and also 4th field in the `SynExpr.Lambda`.  It makes sense to only pass the `SynPat list` and `SynExpr` when creating the lambda just as the parser does.  There is also the case that in a nested lambda the sympathy list in none in the inner expressions and the sequence field is set to true to indicate it is a generated lambda sequence.  All of these details are handled by the compiler so it makes sense to use this function rather than manually mapping `SynPats`.

**Comment by dsyme** [comment]
The current design of FCS does not include helpers for creating syntax tree nodes.   Given one of our goals is to stabilise the FCS API and make it binary compatible I don't think we'll be adding this in the near future.

**Comment by dsyme** [comment]
I'll convert this to a discussion
